providers/implementations/kdfs/argon2.c

```diff
@@ -0,0 +1,1488 @@
+}
+
+static void finalize(const ARGON2_CTX *ctx)
+{
+    block blockhash;
+    uint8_t blockhash_bytes[ARGON2_BLOCK_SIZE];
+    uint32_t last_block_in_lane;
+    uint32_t l;
+
+    if (ctx == NULL)
+        return;
+
+    copy_block(&blockhash, ctx->memory + ctx->lane_length - 1);
+
+    /* XOR the last blocks */
+    for (l = 1; l < ctx->lanes; ++l) {
+        last_block_in_lane = l * ctx->lane_length + (ctx->lane_length - 1);
+        xor_block(&blockhash, ctx->memory + last_block_in_lane);
+    }
+
+    /* Hash the result */
+    store_block(blockhash_bytes, &blockhash);
+    blake2b_long(ctx->out, ctx->outlen, blockhash_bytes, ARGON2_BLOCK_SIZE);
+    OPENSSL_cleanse(blockhash.v, ARGON2_BLOCK_SIZE);
+    OPENSSL_cleanse(blockhash_bytes, ARGON2_BLOCK_SIZE);
+
+    if (ctx->type != ARGON2_D)
+        OPENSSL_secure_clear_free(ctx->memory, ctx->memory_blocks*sizeof(block));
+    else
+        OPENSSL_clear_free(ctx->memory, ctx->memory_blocks*sizeof(block));
+}
+
+static int blake2b(void *out, size_t outlen, const void *in, size_t inlen,
+                   const void *key, size_t keylen)
+{
+    int ret;
+    EVP_MAC *mac;
+    EVP_MAC_CTX *ctx;
+    OSSL_PARAM par[3];
+    size_t par_n, out_written;
+
+    mac = NULL;
+    ctx = NULL;
+    ret = 0;
+    par_n = 0;
+
+    if (out == NULL || outlen == 0)
+        goto fail;
+
+    if ((mac = EVP_MAC_fetch(NULL, "blake2bmac", NULL)) == NULL)
+        goto fail;
+
+    if ((ctx = EVP_MAC_new_ctx(mac)) == NULL)
+        goto fail;
+
+    if (key != NULL && keylen != 0)
+        par[par_n++] = OSSL_PARAM_construct_octet_string(OSSL_MAC_PARAM_KEY,
+                        (void *) key, keylen);
+    par[par_n++] = OSSL_PARAM_construct_size_t(OSSL_MAC_PARAM_SIZE, &outlen);
+    par[par_n++] = OSSL_PARAM_construct_end();
+
+    if (EVP_MAC_set_ctx_params(ctx, par) != 1)
+        goto fail;
+
+    if (EVP_MAC_init(ctx) != 1)
+        goto fail;
+
+    if (EVP_MAC_update(ctx, in, inlen) != 1)
+        goto fail;
+
+    if (EVP_MAC_final(ctx, out, (size_t *) &out_written, outlen) != 1)
+        goto fail;
+
+    ret = 1;
+
+fail:
+    EVP_MAC_free_ctx(ctx);
+    EVP_MAC_free(mac);
+    return ret;
+}
+
+static int blake2b_long(unsigned char *out, size_t outlen, const void *in,
+                        size_t inlen)
+{
+    int ret;
+    EVP_MAC *mac;
+    EVP_MAC_CTX *ctx;
+    OSSL_PARAM par[2];
+    size_t par_n, out_written, blake_outlen;
+    uint32_t outlen_curr;
+    uint8_t outbuf[BLAKE2B_OUTBYTES];
+    uint8_t inbuf[BLAKE2B_OUTBYTES];
+    uint8_t outlen_bytes[sizeof(uint32_t)] = {0};
+
+    ctx = NULL;
+    mac = NULL;
+    ret = 0;
+    par_n = 0;
+    blake_outlen = outlen;
+
+    if (outlen > UINT32_MAX || out == NULL || outlen == 0)
+        goto fail;
+
+    if (outlen > BLAKE2B_OUTBYTES)
+        blake_outlen = BLAKE2B_OUTBYTES;
+
+    /* Ensure little-endian byte order */
+    store32(outlen_bytes, (uint32_t)outlen);
+
+    if ((mac = EVP_MAC_fetch(NULL, "blake2bmac", NULL)) == NULL)
```

Comment:
You might be better off doing this fetch once and then just passing the mac pointer about..

Either way the EVP_MAC_fetch() needs to be passed the libctx anywhere where it is used.

Reply:
Passing mac around in ctx.